doc/specs/#140 - ZIP Support.md

```diff
@@ -0,0 +1,90 @@
+
+> Since we are utilizing Windows Shell APIs, we need to ensure that we are not invoking a UI during the extraction. This can be done by [setting the operation flag to not display any UI](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifileoperation-setoperationflags). 
+
+> During implementation, we will also need to ensure that this process can work under SYSTEM context.
+
+## ZIP Threat Detection
+ZIP and other archive file types are known threat vectors for malware in the form of ZIP compression bombs. Two of the most common types of compression bombs are multi-layered (recursive) and single-layered (non-recursive). ZIP bombs rely on a repetition of identical files that have extremely large compression ratios. 
+
+1. **Multi-layered (recursive)**:
+A ZIP file containing multiple layers of nested ZIP files that achieve high compression ratios when decompressed recursively. This technique is often used to bypass compression ratio checks performed by ZIP parsers for a single layer.
+
+2. **Single-Layered (non-recursive)**
+A ZIP bomb that expands fully after a single round of decompression. The extremely high compression ratio of the ZIP bomb is achieved by overlapping files within the ZIP container. 
+
+In order to protect our users from these possible threats, we will need to implement several basic checks to verify whether it is safe to proceed with the contents of the ZIP archive file.
+
+>Compression Ratio is defined as the sum of the sizes of all the items contained within a ZIP file, divided by the size of the ZIP file.
+
+The check will be as follows:
+
+1. For any archive file, we will start by using Windows Shell APIs which treat ZIP files as a Folder object that can be utilized for viewing the contents (we are not extracting to a specified location, yet). We will iterate through each item and determine the cumulative sum of the sizes of all items contained in the zip and compare that value with the size of the ZIP file. **The check will verify that the compression ratio is no greater than 5:1 (20%)**. Any file that has a larger compression ratio (<20%) will be deemed as "suspicious" and a warning will be displayed to the user. Lossless compression of common data files tend to achieve compression ratios of ~2:1 (50%). Any extraction or copying of files will not take place until this check is passed.
+
+2. If a ZIP archive file contains nested archive files, we will support decompressing up to a maximum of 3 layers (2 additional layers including the initial ZIP file). 
+
+If either of these checks fail, the process will terminate and a warning will be displayed to the user. The user can include `--force` to bypass this check and continue installation at their own risk.

```

Comment:
Why solely 3 layers? That decision appears arbitrary.

Reply:
> Why solely 3 layers? That decision appears arbitrary.

Although it is likely "arbitrary", it is actually an embodiment of good practices. While there are times that putting zipped files inside of zipped files does make sense, it is extremely rare that it is necessary to go more than one layer beyond that. In fact, in [this thread](https://stackoverflow.com/questions/1166385/how-many-times-can-a-file-be-compressed) Nosredna talks about how compressing a file too many times actually leads to the archive being bigger than it originally was.  

So a first layer of compression, which is putting a file inside of a zip, makes a lot of sense. In some cases, a second layer of compression, which is putting a zip inside of a zip, also makes sense, especially when files have a lot of duplicate data. From there, it doesn't really save much extra space to go to that third layer of compression. It can be useful in some situations, but doesn't generally do much. Fourth, fifth, and even more layers of compression are actually more likely to be counterproductive than anything.

Because one layer is common, two layers is not unheard of, and three layers is rare, it doesn't make sense for winget to support more than 3 layers. Additionally, supporting a maximum of 3 layers will force developers to not go overboard on the compression if they want their package to be available in winget

Reply:
@BEEDELLROKEJULIANLOCKHART, it is relatively arbitrary. I decided to opt for a conservative number as it doesn't really make sense to support more than a 3rd layer as @Trenly has explained. If the community strongly feels that 3 is too restrictive, we can increase it but for now I feel that 3 is sufficient.

Reply:
Updated suggested argument for overriding zip malware security checks. We also decided to leverage a zip file malware checker for scanning zip files rather than attempting to write our own.